CLI: Introduced `identity` command to fetch tls-certificates for a pod

jimil749 opened the pull request:
CLI: Introduced `identity` command to fetch tls-certificates for a pod (#4459)


Modified and added a new cli command, which initiates a sni-tls session to the proxy's admin port and returns the certificate.

Usage:
- `linkerd identity pod/<pod-name>`   : fetches certificate from the specified pod
- `linkerd identity -n emojivoto`           : fetches certificate from all pods in the emojivoto namespace
- `linkerd identity -l app=svc/emoji`      : fetches certificate from all pods with label app=svc/emoji

Currently, I am not a 100% sure of the way I am outputting the certificates, I'll change that if needed, but overall these commands do fetch the required certificates.

Screenshot:

![Screenshot from 2020-12-18 22-11-28](https://user-images.githubusercontent.com/47107987/102638794-13466f80-417e-11eb-8861-3f709ab444f9.png)

I would be happy to work upon any changes suggested

adleong commented:
This is so awesome. 😱 

Can't wait to test this out

jimil749 commented:
@adleong, did you get the chance to review/test this?

cpretzer commented:
@jimil749 I had a chance to run some basic tests and this looks good! Thanks for your hard work.

One thing that I noticed is that the command doesn't write any output when Linkerd isn't installed in the cluster.

I had a new kind cluster running without Linkerd installed, and when I ran the `linkerd identity ...` command, the command returned with no output or indication that anything happened.

In contrast, the `linkerd stat deploy` and `linkerd edges deploy` commands have this output when the Linkerd control plane is not deployed:

```bash
$ linkerd stat deploy
Cannot find Linkerd: configmaps "linkerd-config" not found
Validate the install with: linkerd check
```

What do you think about adding a check to make sure that linkerd is installed?

jimil749 commented:
@cpretzer, thanks for the feedback, I'll do the needful. :)

cpretzer commented:
👍

jimil749 commented:
@cpretzer I think the reason we aren't getting any output is because we do not have the proxy sidecar container injected. The `metrics` command behaves the same way as `identity` (returns with no output or any indication)! I am not using the control plane APIs to fetch the certs, I am simply establishing a sni-tls connection to the proxy's admin port to do so. The check for whether the control plane is installed or not is done while building a new public API client (in case of `stat` and `edges`), which is not required in my case. I am currently not a 100% sure of how to add the checks. I think checking whether the service is meshed i.e the proxy exists would be a better check.

cpretzer commented:
@jimil749 I see what you're saying. The `edges` and `stats` commands use the `checkPublicAPIClientOrExit` [function](https://github.com/linkerd/linkerd2/blob/main/cli/cmd/edges.go#L85)

Perhaps the correct approach is similar to the one used by the `uninstall` [command](https://github.com/linkerd/linkerd2/blob/main/cli/cmd/uninstall.go#L41), which checks whether the linkerd namespace exists and exits if it does not.

jimil749 commented:
@cpretzer, Sorry for the delayed response, but I've added the check for the linkerd control plane but I wasn't able to test the command entirely, because I am having an error in installing the control plane in my cluster after fetching the latest commit, which I think is because of [this](https://github.com/linkerd/linkerd2/pull/5340), I am getting the same error as in this [comment](https://github.com/linkerd/linkerd2/pull/5340#issuecomment-747104433). Not a 100% sure, but is that the reason the tests are failing?

cpretzer commented:
hi @jimil749 thanks for adding the check for the namespace. We're working on getting the build fixed, and once it is, we can run the tests again.

adleong reviewed (commented):
I played with this a little bit and it's pretty great.

I like the `linkerd identity (pod-name)` mode, though I think we should drop the resource type or at least make it optional since only pods are supported.  i.e. instead of `linkerd identity pod/linkerd-destination-6887bb8f49-h8qvr` it should be `linkerd identity linkerd-destination-6887bb8f49-h8qvr`

I also like the `-l` selector mode.

I found the third mode where it prints all certificates in a namespace to be confusing and I think we can remove that.

I also looked into certificate formatting a bit more and found https://github.com/smallstep/certinfo.  I think we should use this library for formatting the certificates.

adleong commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,295 @@
+	var activeRoutines int32
+	for _, pod := range pods {
+		atomic.AddInt32(&activeRoutines, 1)
+		go func(p corev1.Pod) {
```
I think it's fine to fetch each certificate in series rather than doing them in parallel.  This should let us simplify this and get rid of the channel.  It also happens to fix the bug where if there are no pods selected by the label selector, the select hangs.

adleong commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,295 @@
+	re := regexp.MustCompile("[0-9]+")
+	res := re.FindString(url)
+
+	connURL := "localhost:" + res
```
Can you explain what's going on here?  We're extracting the port from the port-forward url and appending it onto localhost?  Isn't the port-forward url already a localhost address?

jimil749 replied:
The thing is that the `tls.Dial` function in Go expects a domain name with port and not the entire url, and `portForward.URLFor()` gives us the URL. Just had to trim the `http://` part from the url, which I now did using the `strings.Trim`. I agree this was an overkill and changed it.

adleong commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,295 @@
+	roots.AppendCertsFromPEM([]byte(rootPEM))
+
+	conn, err := tls.Dial("tcp", connURL, &tls.Config{
+		RootCAs:    roots,
```
Do we actually need to specify the rootCA and the serverName here?  Can we use `InsecureSkipVerify` instead?  We just want to print the certificate, not validate it.

jimil749 replied:
We need the serverName here, but yes if we use `InsecureSkipVerify` we can do away the the rootCA.

jimil749 commented:
> 
> I like the `linkerd identity (pod-name)` mode, though I think we should drop the resource type or at least make it optional since only pods are supported. i.e. instead of `linkerd identity pod/linkerd-destination-6887bb8f49-h8qvr` it should be `linkerd identity linkerd-destination-6887bb8f49-h8qvr`
> 
Yeah, makes sense. Will do the necessary changes!
> 
> I found the third mode where it prints all certificates in a namespace to be confusing and I think we can remove that.
> 
The reason to do was this [comment](https://github.com/linkerd/linkerd2/issues/4459#issuecomment-684084431), where it is mentioned that a command like `linkerd -n ns identity` can be used to get certificates for all pods in the namespace. I will remove the command if it is not required as such
> I also looked into certificate formatting a bit more and found https://github.com/smallstep/certinfo. I think we should use this library for formatting the certificates.
>
I will take a look at the library. :+1:

jimil749 commented:
@adleong, I've addressed the changes proposed, here's the summary of the changes:
- Fetching the certificates in series rather than parallel.
- URL trim using `strings.Trim` for `tls.Dial()` (instead of extracting the portno from the URL)
- Dropped the resource type from `linkerd identity` command
- Removed the `linkerd -n ns` subcommand, which used to fetch certificates for all pods in the provided ns
- Used `grantae/certinfo` library to print the certificates in a format similar to OpenSSL
- Removed RootCAs field from `tls.Dial`

adleong commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,231 @@
+				return err
+			}
+
+			resources, err := resource.FetchKubernetesResources(cmd.Context(), k8sAPI, metav1.ListOptions{LabelSelector: k8s.ControllerNSLabel})
```
I don't think we need this check that the control plane exists.

adleong commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,231 @@
+					return nil
+				}
+				fmt.Print(result)
+				fmt.Println("pem: |")
```
Since we're already printing the entire certificate in an openssl-like format, I don't think we also need to print the PEM encoding.

adleong commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,231 @@
+			}
+
+			resultCerts := getCertificate(k8sAPI, pods, k8s.ProxyAdminPortName, emitLog)
+			for _, resultCert := range resultCerts {
```
Can we restore the feature where, if printing multiple certificates, it shows the pod name?

adleong commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,231 @@
+				return err
+			}
+
+			resultCerts := getCertificate(k8sAPI, pods, k8s.ProxyAdminPortName, emitLog)
```
We should probably print an error message if no certificate are returned here.  i.e. if the user specified an un-meshed pod.

adleong reviewed (commented):
This is so awesome!  I'm really excited for this to land. 😻

jimil749 commented:
@adleong, I made the changes as suggested. But I was not sure of what exact error message to print out when no certificates are returned. Maybe you could suggest a better message to output?

kleimkuhler reviewed (commented):
@jimil749 This is super cool thanks for getting this change up!

I've left a few smaller comments focused around formatting help text and such.

There are a few comments though that are more significant. Let me know if you have any questions about those I'd be glad to clarify a little more!

jimil749 commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,224 @@
+}
+
+func getPods(ctx context.Context, clientset kubernetes.Interface, namespace string, selector string, arg []string) ([]corev1.Pod, error) {
+	if len(arg) > 0 {
+		pod, err := clientset.CoreV1().Pods(namespace).Get(ctx, arg[0], metav1.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+		return []corev1.Pod{*pod}, nil
+	}
```
Oh alright. To be honest, I had not thought about such a situation, I was under the impression that we just want to fetch cert for a single pod. I will make the required changes! :)

jimil749 commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,224 @@
+}
+
+func getCertResponse(url string, pod corev1.Pod) ([]*x509.Certificate, error) {
+	serverName, err := getServerName(pod, "linkerd-proxy")
```
Just to be sure, I am supposed to add the const in `linkerd2/pkg/k8s/labels.go`, right?

adleong commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,224 @@
+
+			resultCerts := getCertificate(k8sAPI, pods, k8s.ProxyAdminPortName, emitLog)
+			if len(resultCerts) == 0 {
+				fmt.Print("Could not fetch Certificate. Ensure that the pod(s) are meshed by running `linkerd inject`")
```
I think this error message is good!  It does need to have a newline at the end though.

adleong reviewed (approved):
Thanks for updating, @jimil749!

jimil749 commented:
@kleimkuhler, I've made the required changes!

kleimkuhler commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,227 @@
+			for i, resultCert := range resultCerts {
+				fmt.Printf("\nPOD %s (%d of %d)\n\n", resultCert.pod, i+1, len(resultCerts))
+				if resultCert.err != nil {
+					fmt.Printf("\n%s", resultCert.err)
```
We should have a newline here after printing the error message or else (probably depending on your terminal) it indicate that it's missing. You can see that symbol at the end of output below:

```
❯ ./target/release/linkerd2-cli-git-a3b2afa6-linux-amd64 identity -n kube-system local-path-provisioner-6d59f47c7-7fnt7

POD local-path-provisioner-6d59f47c7-7fnt7 (1 of 1)


failed to find linkerd-admin port in linkerd-proxy container for given pod spec⏎ 
```

```suggestion
					fmt.Printf("\n%s\n", resultCert.err)
```

kleimkuhler commented on cli/cmd/identity.go:
```diff
@@ -0,0 +1,227 @@
+				cert := certChain[len(certChain)-1]
+				result, err := certinfo.CertificateText(cert)
+				if err != nil {
+					fmt.Printf("\n%s", err)
```
Here as well!

kleimkuhler reviewed (commented):
This looks great! I have one more suggestion to just make sure errors are printing as expected.

After that it should be good. Sorry I did not catch this in the previous review!

kleimkuhler commented:
Thanks for addressing all these reviews @jimil749!
